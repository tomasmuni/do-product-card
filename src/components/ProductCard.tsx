import React, { createContext } from 'react';

import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, OnChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/product.interfaces';

import styles from '../styles/styles.module.css'

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;



export interface Props {
    product: Product;
    // children?: React.ReactElement | React.ReactElement[];
    children: (args: ProductCardHandlers) => React.ReactElement
    className?: string;
    style?: React.CSSProperties;
    onChange?: ( args: OnChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues
}


export const ProductCard = ({ children, product, className, style, onChange, value , initialValues}: Props ) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount
        }}>
            <div 
                className={ `${ styles.productCard } ${ className }` }
                style={ style }
            >
                { children({
                    count: counter,
                    product,
                    isMaxCountReached,
                    increaseBy,
                    reset,
                    maxCount: initialValues?.count
                }) }
            </div>
        </Provider>
    )
}