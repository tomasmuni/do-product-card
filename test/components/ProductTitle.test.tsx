import React from 'react'
import renderer from 'react-test-renderer'
import { ProductTitle, ProductCard } from '../../src/components';
import { product2 } from '../data/products';



describe('ProductTitle',() => {

    test("debe mostrar el componente con el titulos", () => {
        const wrapper = renderer.create(
            <ProductTitle title='Custom title' className='custom-class'></ProductTitle>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    })

    test("debe mostrar el componente con el nombre del producto", () => {
        const wrapper = renderer.create(
            <ProductCard product={product2}> 
            {
                () => (
                    <ProductTitle></ProductTitle>
                )

            } 
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    })

})