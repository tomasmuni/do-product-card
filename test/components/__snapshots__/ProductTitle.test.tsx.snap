// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Card Sin Imagen
  </span>
</div>
`;

exports[`ProductTitle debe mostrar el componente con el titulos 1`] = `
<span
  className="productDescription custom-class"
>
  Custom title
</span>
`;
